File: DreamTeam/src/main/java/team/project/redboost/entities/Projet.java
package team.project.redboost.entities;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.fasterxml.jackson.datatype.jsr310.deser.LocalDateDeserializer;
import com.fasterxml.jackson.datatype.jsr310.ser.LocalDateSerializer;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
@Entity
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@JsonIgnoreProperties({"entrepreneurs", "produits", "services", "founder", "pendingCollaborator"})
public class Projet {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String name;
    private String sector;
    private String type;
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
    @JsonDeserialize(using = LocalDateDeserializer.class)
    @JsonSerialize(using = LocalDateSerializer.class)
    private LocalDate creationDate;
    private String description;

    @Enumerated(EnumType.STRING)
    private Objectives objectives;

    @Enumerated(EnumType.STRING)
    private Statut status;

    private Double globalScore;
    private String location;

    @Column(length = 500)
    private String logoUrl;

    private String websiteUrl;
    private Double revenue;
    private Long numberOfEmployees;
    private Long nbFemaleEmployees;
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
    @JsonDeserialize(using = LocalDateDeserializer.class)
    @JsonSerialize(using = LocalDateSerializer.class)
    private LocalDate lastUpdated;
    private String associatedSectors;
    private String technologiesUsed;
    private Double fundingGoal;
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
    @JsonDeserialize(using = LocalDateDeserializer.class)
    @JsonSerialize(using = LocalDateSerializer.class)
    private LocalDate lastEvaluationDate;

    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
    @JoinColumn(name = "projet_id")
    private List<Produit> produits = new ArrayList<>();

    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
    @JoinColumn(name = "projet_id")
    private List<ServiceP> services = new ArrayList<>();

    @ManyToMany
    @JoinTable(
            name = "projet_entrepreneur",
            joinColumns = @JoinColumn(name = "projet_id"),
            inverseJoinColumns = @JoinColumn(name = "user_id")
    )

    @JsonIgnoreProperties("projets")
    private List<User> entrepreneurs = new ArrayList<>();

    @ManyToMany
    @JoinTable(
            name = "projet_coach",
            joinColumns = @JoinColumn(name = "projet_id"),
            inverseJoinColumns = @JoinColumn(name = "user_id")
    )
    private List<User> coaches = new ArrayList<>();

    @ManyToMany
    @JoinTable(
            name = "projet_investor",
            joinColumns = @JoinColumn(name = "projet_id"),
            inverseJoinColumns = @JoinColumn(name = "user_id")
    )
    private List<User> investors = new ArrayList<>();

    @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE})
    @JoinTable(
            name = "projet_folder",
            joinColumns = @JoinColumn(name = "projet_id"),
            inverseJoinColumns = @JoinColumn(name = "folder_id")
    )
    private List<FolderMetadata> folders = new ArrayList<>();

    @OneToMany(mappedBy = "projet", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Phase> phases = new ArrayList<>();

    // Track the initiator (owner) of the project
    @ManyToOne
    @JoinColumn(name = "founder_id", nullable = false)
    @JsonIgnoreProperties("projets")
    private User founder;

    // Optional: Track a pending collaborator (nullable, set only when inviting)
    @ManyToOne
    @JoinColumn(name = "pending_collaborator_id")
    @JsonIgnoreProperties("projets")
    private User pendingCollaborator;

    public enum Statut {
        EN_DEVELOPPEMENT, OPERATIONNELLE, EN_RECHERCHE_FINANCEMENT, TERMINE
    }

    public enum Objectives {
        COURT_TERME, MOYEN_TERME, LONG_TERME
    }
}

File: Redboost_front/src/app/models/Projet.ts
export enum Objectives {
    COURT_TERME = 'COURT_TERME',
    MOYEN_TERME = 'MOYEN_TERME',
    LONG_TERME = 'LONG_TERME'
}

export enum Statut {
    EN_DEVELOPPEMENT = 'EN_DEVELOPPEMENT',
    OPERATIONNELLE = 'OPERATIONNELLE',
    EN_RECHERCHE_FINANCEMENT = 'EN_RECHERCHE_FINANCEMENT',
    TERMINE = 'TERMINE'
}
import { Produit } from './Produit';
import { ServiceP } from './ServiceP';

export class Projet {
    id?: number;
    name: string;
    sector: string;
    type: string;
    creationDate: string;
    description: string;
    objectives: Objectives;
    status: Statut;
    globalScore: number;
    location: string;
    logoUrl: string;
    websiteUrl: string;
    revenue: number;
    numberOfEmployees: number;
    nbFemaleEmployees: number;
    lastUpdated: string;
    associatedSectors: string;
    technologiesUsed: string;
    fundingGoal: number;
    lastEvaluationDate: string;
    produits: Produit[];
    services: ServiceP[];
    folders: any[];
    phases: any[];
    founder: number | null; // Changed from User to number
    entrepreneurs: number[]; // Changed from User[] to number[]
    coaches: number[]; // Changed from User[] to number[]
    investors: number[]; // Changed from User[] to number[]
    pendingCollaborator: number | null; // Changed from User to number

    constructor(
        name: string,
        sector: string,
        type: string,
        creationDate: string,
        description: string = '',
        objectives: Objectives = Objectives.COURT_TERME,
        status: Statut = Statut.EN_DEVELOPPEMENT,
        globalScore: number = 0,
        location: string = '',
        logoUrl: string = '',
        websiteUrl: string = '',
        revenue: number = 0,
        numberOfEmployees: number = 0,
        nbFemaleEmployees: number = 0,
        lastUpdated: string = '',
        associatedSectors: string = '',
        technologiesUsed: string = '',
        fundingGoal: number = 0,
        lastEvaluationDate: string = '',
        produits: Produit[] = [],
        services: ServiceP[] = [],
        folders: any[] = [],
        phases: any[] = [],
        founder: number | null = null, // Changed from User to number
        entrepreneurs: number[] = [], // Changed from User[] to number[]
        coaches: number[] = [], // Changed from User[] to number[]
        investors: number[] = [], // Changed from User[] to number[]
        pendingCollaborator: number | null = null // Changed from User to number
    ) {
        this.id = undefined;
        this.name = name;
        this.sector = sector;
        this.type = type;
        this.creationDate = creationDate;
        this.description = description;
        this.objectives = objectives;
        this.status = status;
        this.globalScore = globalScore;
        this.location = location;
        this.logoUrl = logoUrl;
        this.websiteUrl = websiteUrl;
        this.revenue = revenue;
        this.numberOfEmployees = numberOfEmployees;
        this.nbFemaleEmployees = nbFemaleEmployees;
        this.lastUpdated = lastUpdated;
        this.associatedSectors = associatedSectors;
        this.technologiesUsed = technologiesUsed;
        this.fundingGoal = fundingGoal;
        this.lastEvaluationDate = lastEvaluationDate;
        this.produits = produits;
        this.services = services;
        this.folders = folders;
        this.phases = phases;
        this.founder = founder;
        this.entrepreneurs = entrepreneurs;
        this.coaches = coaches;
        this.investors = investors;
        this.pendingCollaborator = pendingCollaborator;
    }
}
